File: apps/dsa-web/src/api/usage.ts
import apiClient from './index';
import { toCamelCase } from './utils';

export interface CallTypeBreakdown {
  callType: string;
  calls: number;
  totalTokens: number;
}

export interface ModelBreakdown {
  model: string;
  calls: number;
  totalTokens: number;
}

export interface UsageSummary {
  period: string;
  fromDate: string;
  toDate: string;
  totalCalls: number;
  totalTokens: number;
  byCallType: CallTypeBreakdown[];
  byModel: ModelBreakdown[];
}

export type UsagePeriod = 'today' | 'month' | 'all';

export const usageApi = {
  async getSummary(period: UsagePeriod = 'month'): Promise<UsageSummary> {
    const res = await apiClient.get('/api/v1/usage/summary', { params: { period } });
    return toCamelCase<UsageSummary>(res.data);
  },
};

File: api/v1/schemas/usage.py
# -*- coding: utf-8 -*-
"""Schemas for LLM usage tracking API."""

from __future__ import annotations

from typing import List, Optional

from pydantic import BaseModel, Field


class CallTypeBreakdown(BaseModel):
    call_type: str = Field(..., description="'analysis' | 'agent' | 'market_review'")
    calls: int
    prompt_tokens: int = 0
    completion_tokens: int = 0
    total_tokens: int


class ModelBreakdown(BaseModel):
    model: str
    provider: Optional[str] = None
    calls: int
    prompt_tokens: int = 0
    completion_tokens: int = 0
    total_tokens: int
    max_total_tokens: int = 0
    context_window: Optional[int] = None
    context_usage_ratio: Optional[float] = Field(
        default=None,
        description="total_tokens / context_window when model metadata is available",
    )


class UsageCallRecord(BaseModel):
    id: int
    called_at: str = Field(..., description="ISO datetime string")
    call_type: str
    model: str
    provider: Optional[str] = None
    stock_code: Optional[str] = None
    prompt_tokens: int
    completion_tokens: int
    total_tokens: int
    context_window: Optional[int] = None
    context_usage_ratio: Optional[float] = None


class UsageSummaryResponse(BaseModel):
    period: str = Field(..., description="'today' | 'month' | 'all'")
    from_date: str = Field(..., description="ISO date string")
    to_date: str = Field(..., description="ISO date string")
    total_calls: int
    total_prompt_tokens: int = 0
    total_completion_tokens: int = 0
    total_tokens: int
    by_call_type: List[CallTypeBreakdown]
    by_model: List[ModelBreakdown]


class UsageDashboardResponse(UsageSummaryResponse):
    recent_calls: List[UsageCallRecord]
